synapse/handlers/send_email.py

```diff
@@ -106,7 +107,10 @@ def build_sender_factory(**kwargs) -> ESMTPSenderFactory:
         factory = build_sender_factory(hostname=smtphost if enable_tls else None)
 
     # the IReactorTCP interface claims host has to be a bytes, which seems to be wrong
-    reactor.connectTCP(smtphost, smtpport, factory, timeout=30, bindAddress=None)  # type: ignore[arg-type]
+    endpoint = HostnameEndpoint(
+        reactor, smtphost, smtpport, timeout=30, bindAddress=None
+    )
+    endpoint.connect(factory)
```

Comment:
As @clokep said, I think we need to `await` this. In fact, for logcontext-related reasons too annoying to explain here, we need to:

```suggestion
    await make_deferred_yieldable(endpoint.connect(factory))
```